src/pybind/mgr/cephadm/serve.py

```diff
@@ -845,8 +855,24 @@ def has_interface_for_vip(host: str, sspec: ServiceSpec) -> bool:
                                         f"Failed to apply {len(self.mgr.apply_spec_fails)} service(s): {','.join(x[0] for x in self.mgr.apply_spec_fails)}",
                                         len(self.mgr.apply_spec_fails),
                                         warnings)
-            return False
+            raise OrchestratorError(msg)
 
+        return all_slots, slots_to_add, daemons_to_remove
+
+    def deploy_and_remove_daemons_by_service(
+        self,
+        all_slots: List[DaemonPlacement],
+        slots_to_add: List[DaemonPlacement],
+        daemons_to_remove: List[orchestrator.DaemonDescription],
```

Comment:
Regarding my earlier comment - imagine this looked like:
```
def deploy_and_remove_daemons_by_service(self, changes: ChangeGroup) -> bool:
```
how much more compact and information dense it becomes and permits extensibility in the future.